app/models/concerns/course_component_query_concern.rb

```diff
@@ -0,0 +1,78 @@
+# frozen_string_literal: true
+module CourseComponentQueryConcern
+  extend ActiveSupport::Concern
+
+  # @return [Array<Class>] The classes of the components that are available
+  def available_components
+    raise NotImplementedError, 'Concrete concern must implement available_components'
+  end
+
+  # @return [Array<Class>] The subset of available_components that the user can disable.
+  def disableable_components
+    raise NotImplementedError, 'Concrete concern must implement disableable_components'
+  end
+
+  def undisableable_components
+    @undisableable_components ||= available_components - disableable_components
+  end
+
+  # Applies user preferences to components that can be disabled.
+  #
+  # @return [Array<Class>] Array of components that are effectively enabled.
+  def enabled_components
+    @enabled_components ||= undisableable_components | user_enabled_components
+  end
+
+  # @return [Array<Class>] Components specified as 'enabled' by the user.
+  def user_enabled_components
+    @user_enabled_components ||= available_components.select do |component|
+      enabled = component_setting(component.key).enabled
+      enabled.nil? ? component.enabled_by_default? : enabled
+    end
+  end
+
+  # Set component's `enabled` key to be either true or false.
+  #
+  # @param [Symbol|String] key Component key
+  # @param [Boolean] value true if component is to be enabled, false otherwise.
+  def unsafe_set_component_enabled_boolean(key, value)
```

Comment:
1 more comment: My preference would be not to abstract out this method, or keep it as a private method. The method name is too long as compared to the actual operation, which is quite readable. Can I suggest we just do away with this method, unless I am missing a use case?

Reply:
This is used in the specs. Having this method would allow the spec not to have to know which slice of the settings hash to persist the course component settings in.

Reply:
Would it be better to add this method to the specs then, rather than including it here?

Reply:
I would prefer not to because then I would have to expose `#component_setting` and let the specs handle the settings_on_rails settings slice directly.

Although its name is verbose, I do think that using the method helps to highlight to the user that s/he is doing something that circumvents the normal (safe) flow. Similar spirit to `dangerouslySetInnerHTML`.

It is worth it trade off the verbosity for clarity (@fonglh's feedback) and the abstraction barrier imho.

Reply:
Or how about as a private method instead? I do use `object.send(:method)` in certain times for specs when there are dependencies. The end goal of public methods is for reuse in other parts of the app, but not necessarily for specs. 

It does seem similar to `dangeroulySetInnerHTML` but that is a separate workaround for a library used by many external parties, whereas in our case I do not expect the need for this function in other parts of the app.

Reply:
ok, I'll make it a private method and use `send`.